src/hotspot/cpu/x86/macroAssembler_x86_poly.cpp

```diff
@@ -0,0 +1,874 @@
+  vpaddq(A0, A0, T0, Assembler::AVX_128bit);
+  vpaddq(A1, A1, T1, Assembler::AVX_128bit);
+  vpaddq(A2, A2, T2, Assembler::AVX_128bit);
+
+  // 2 -> 1 blocks
+  vpsrldq(T0, A0, 8, Assembler::AVX_128bit);
+  vpsrldq(T1, A1, 8, Assembler::AVX_128bit);
+  vpsrldq(T2, A2, 8, Assembler::AVX_128bit);
+
+  // Finish folding and clear second qword
+  mov64(t0, 0xfd);
+  kmovql(k1, t0);
+  evpaddq(A0, k1, A0, T0, false, Assembler::AVX_512bit);
+  evpaddq(A1, k1, A1, T1, false, Assembler::AVX_512bit);
+  evpaddq(A2, k1, A2, T2, false, Assembler::AVX_512bit);
+
+  // Carry propagation
+  vpsrlq(T0, A0, 44, Assembler::AVX_512bit);
+  evpandq(A0, A0, Address(polyCP, mask_44), Assembler::AVX_512bit); // Clear top 20 bits
+  vpaddq(A1, A1, T0, Assembler::AVX_512bit);
+  vpsrlq(T0, A1, 44, Assembler::AVX_512bit);
+  evpandq(A1, A1, Address(polyCP, mask_44), Assembler::AVX_512bit); // Clear top 20 bits
+  vpaddq(A2, A2, T0, Assembler::AVX_512bit);
+  vpsrlq(T0, A2, 42, Assembler::AVX_512bit);
+  evpandq(A2, A2, Address(polyCP, mask_42), Assembler::AVX_512bit); // Clear top 22 bits
+  vpsllq(T1, T0, 2, Assembler::AVX_512bit);
+  vpaddq(T0, T0, T1, Assembler::AVX_512bit);
+  vpaddq(A0, A0, T0, Assembler::AVX_512bit);
+
+  // Put together A (accumulator)
+  movq(a0, A0);
+
+  movq(t0, A1);
+  movq(t1, t0);
+  shlq(t1, 44);
+  orq(a0, t1);
+
+  shrq(t0, 20);
+  movq(a2, A2);
+  movq(a1, a2);
+  shlq(a1, 24);
+  orq(a1, t0);
+  shrq(a2, 40);
+
+  // Cleanup
+  vpxorq(xmm0, xmm0, xmm0, Assembler::AVX_512bit);
+  vpxorq(xmm1, xmm1, xmm1, Assembler::AVX_512bit);
+  vpxorq(T0, T0, T0, Assembler::AVX_512bit);
+  vpxorq(T1, T1, T1, Assembler::AVX_512bit);
+  vpxorq(T2, T2, T2, Assembler::AVX_512bit);
+  vpxorq(C0, C0, C0, Assembler::AVX_512bit);
+  vpxorq(C1, C1, C1, Assembler::AVX_512bit);
+  vpxorq(C2, C2, C2, Assembler::AVX_512bit);
+  vpxorq(A0, A0, A0, Assembler::AVX_512bit);
+  vpxorq(A1, A1, A1, Assembler::AVX_512bit);
+  vpxorq(A2, A2, A2, Assembler::AVX_512bit);
+  vpxorq(A3, A3, A3, Assembler::AVX_512bit);
+  vpxorq(A4, A4, A4, Assembler::AVX_512bit);
+  vpxorq(A5, A5, A5, Assembler::AVX_512bit);
+  vpxorq(B0, B0, B0, Assembler::AVX_512bit);
+  vpxorq(B1, B1, B1, Assembler::AVX_512bit);
+  vpxorq(B2, B2, B2, Assembler::AVX_512bit);
+  vpxorq(B3, B3, B3, Assembler::AVX_512bit);
+  vpxorq(B4, B4, B4, Assembler::AVX_512bit);
+  vpxorq(B5, B5, B5, Assembler::AVX_512bit);
+  vpxorq(R0, R0, R0, Assembler::AVX_512bit);
+  vpxorq(R1, R1, R1, Assembler::AVX_512bit);
+  vpxorq(R2, R2, R2, Assembler::AVX_512bit);
+  vpxorq(R1P, R1P, R1P, Assembler::AVX_512bit);
+  vpxorq(R2P, R2P, R2P, Assembler::AVX_512bit);
+  evmovdquq(A0, Address(rsp, 64*3), Assembler::AVX_512bit);
+  evmovdquq(A0, Address(rsp, 64*4), Assembler::AVX_512bit);
+  evmovdquq(A0, Address(rsp, 64*5), Assembler::AVX_512bit);
+  evmovdquq(A0, Address(rsp, 64*0), Assembler::AVX_512bit);
+  evmovdquq(A0, Address(rsp, 64*1), Assembler::AVX_512bit);
+  evmovdquq(A0, Address(rsp, 64*2), Assembler::AVX_512bit);
```

Comment:
This is load from stack into A0. Did you intend to store A0 (cleanup) into stack local area here? I think the source and destination are mixed up here.

Reply:
Wow! Thank you for spotting this

Reply:
done